src/Adapter/Product/Customization/Update/ProductCustomizationFieldUpdater.php

```diff
@@ -75,30 +75,37 @@ public function __construct(
 
     /**
      * @param ProductId $productId
-     * @param CustomizationField[] $customizationFields
+     * @param LegacyCustomizationField[] $customizationFields
+     * @param ShopConstraint $shopConstraint
      */
-    public function setProductCustomizationFields(ProductId $productId, array $customizationFields): void
-    {
-        $product = $this->productRepository->get($productId);
-        $deletableFieldIds = $this->getDeletableFieldIds($customizationFields, $product);
+    public function setProductCustomizationFields(
+        ProductId $productId,
+        array $customizationFields,
+        ShopConstraint $shopConstraint
+    ): void {
+        $productShops = $this->productRepository->getAssociatedShopIds($productId);
+        $deletableFieldIds = $this->getDeletableFieldIds($customizationFields, $productId);
 
         foreach ($customizationFields as $customizationField) {
             if ($customizationField->id) {
-                $this->customizationFieldRepository->update($customizationField);
+                $this->customizationFieldRepository->update($customizationField, [$shopConstraint->getShopId()]);
             } else {
-                $this->customizationFieldRepository->add($customizationField);
+                $this->customizationFieldRepository->add($customizationField, $productShops);
             }
         }
 
         $this->customizationFieldDeleter->bulkDelete($deletableFieldIds);
-        $this->refreshProductCustomizability($product);
+        $this->refreshProductCustomizability($productId);
     }
 
     /**
-     * @param Product $product
+     * @param ProductId $productId
      */
-    public function refreshProductCustomizability(Product $product): void
+    public function refreshProductCustomizability(ProductId $productId): void
     {
+        // Although the modified fields are defined as multishop, they depend on the association with customization
+        // fields which are not, so all those fields must be synced on all associated shop
```

Comment:
Kind of tthe same, but I think its clearer (was abit hard to understand from first look)
```suggestion
        // The modified fields are defined as multishop, but they depend on the association with customization
        // fields which are not multishop, so all those fields must be synced between all associated shops
```